docs/proposals/innerloop-to-outerloop.md

```diff
@@ -0,0 +1,138 @@
+
+
+
+## Usage of Build and Deploy guidance in CI/CD and Gitops
+
+
+### Summary 
+
+
+`odo push` syncs user's source code to a workspace container  run using the guidance from a runtime-specific `devfile`.
+
+Example, a Sprintboot project, when associated with a Springboot specific devfile, provides `odo`  the know-how to run the inner-loop container with workspace dependencies
+like debug tools and SDKs.
+
+`odo deploy` would read guidance from the devfile on how to build or more images and deploy the same. The devfile would either be a registry-provided devfile based on the user's choice of runtime or a project-specific devfile that was present at the root of the project.
+
+`odo pipelines` would generate configuration and manifests locally on the filesystem to eventually manage the application component from Git using CI/CD pipelines. The manifests are to be generated based on the build and deploy guidance from the devfile.
+
+### The outerloop and Gitops story
+
+
+In the initial iterations of supporting outerloop, the `odo pipelines` command would be used to generate Tekton and Argo manifests for CI and CD, respectively, for enabling users to drive Gitops-based workflows, organized in an opinionated directory structure in Git.
+
+To enable generation of runtime-specific near-accurate pipelines, `odo` would read the image build and deploy guidance from the devfile, the same way `odo deploy` would 
+do an outerloop build and image deployment on the cluster using guidance from the devfile.
+
+#### Build
+As an example, for a devfile associated with an application component, a dockerfile-based image build guidance in the `devfile` would be consumed by the `odo pipelines` command to
+* generate [Tekton Task yaml manifests](https://github.com/tektoncd/catalog/blob/master/task/buildah/0.1/buildah.yaml) which does a Dockerfile build, or 
+* generate an equivalent [k8s-build manifest](https://github.com/redhat-developer/build/blob/master/samples/build/build_buildah_cr.yaml), orchestrated by Tekton Pipelines.
+* directly use a Tekton Task referenced in the `devfile`.
+
+#### Deploy
+Similarly, the `odo pipelines` command would generate 
+* The necessary 'yaml' manifests to deploy the built image using the guidance from the devfile.
+* The Argo CRs that represent apps that would be reconciled by Argo on a given cluster.
+
+The generated manifests may optionally be consumed outside the context of `odo` using plain old `kubectl` and `kustomize`. Overtime, the user may choose to modify them and commit them
+to Git.
+
+## Enhancements to the devfile
+
+The build and deploy guidances would primarily feature as new devfile components in `devfile` v2.1.0.
+
+Example, the current proposal to add a devfile component called `dockerfile` looks like: 
+
+
+```
+schemaVersion: 2.1.0
+metadata:
+  name: nodejs
+  version: 1.0.0
+  alpha.deployment-manifest: "https://raw.githubusercontent.com/groeges/devfile-registry/master/devfiles/nodejs/deploy_deployment.yaml"
```

Comment:
If we are talking about devfile v2.1 shouldn't there be "first-class" support for deployment manifests?

Reply:
I gotta remove this, copy-pasted by mistake. Thanks for pointing out.